### DISCUSSION TITLE
Does plugin "Images " support  to find part image which comes from one full screenshot?

### DISCUSSION THREAD

**Discussion by billychen2014** [appium/appium#18841] [Q&A] [answered]
hello,

"Images" finding  is amazing function , i did some testing against it:

1.  try to find using 'find_element_by_image（**full screenshot png file**）' ==> result is ok
2. try to find using 'find_element_by_image（**part of full screenshot png file**）' ==> result is fail

so, it looks like the image finding only support  full screenshot finding till now? [i also tried the update driver setting such as 'fixImageTemplateScale' etc ]

the picture like below:
<img width="339" alt="image" src="https://github.com/appium/appium/assets/7981655/b2f7af21-ec35-4a06-8e2e-0489528c811a">

**Comment by Higgens75** [comment]
Maybe i missunderstand your meaning of "part of full screenshot" but for me the function works.
 I can find parts of the screen (i.e. a button) with a picture from this button.

driver.findElement(AppiumBy.image(imagePathBase64("button_picture.png")));

**Reply by billychen2014** [reply]
hmm,  
part of full screenshot just like your scenario : i have one whole full screen ,and i captured images from this screen file(one button ) and try to find the button using find_element_by_image.

**Reply by Higgens75** [reply]
i use appium2 with java, maybe there are differences

**Reply by billychen2014** [reply]
aha, i use appium2 with py

**Reply by Higgens75** [reply]
do you capture the part.png within the app or do you snip it manually out of a screenshot ?

Maybe this appiumPro tutorials help a bit. Because it looks like you want to screen compare.

https://appiumpro.com/editions/98-visual-testing-with-appium-part-1
https://appiumpro.com/editions/99-visual-testing-with-appium-part-2
https://appiumpro.com/editions/100-visual-testing-with-appium-part-3

**Reply by billychen2014** [reply]
do you capture the part.png within the app or do you snip it manually out of a screenshot ? ==》manually

i'm trying to download java client and will have a try , i also need to read the guide above carefully firstly, thanks

**Reply by Higgens75** [reply]
think manually is not a good idea. There could be differences in resolution or other stuff when saving the picture with an external tool, which generates differences.

In the first tutorial is a good code snippet, which shows how to capture a complete screen within your test to compare it while the test runs. For doing this for a part of the screen there has to be made some modifications to that code.

**Reply by billychen2014** [reply]
there has to be made some modifications to that code ==> i agree. it looks like that appium V2  did not support such function till now .
if there is one function named 'save_screenshot' with args 'x,y,w,h' ,that would be perfect.

**Reply by Higgens75** [reply]
its something like this in java
public static BufferedImage generateImageByCoords(File screenshot) throws IOException {
	    BufferedImage fullImage = ImageIO.read(screenshot);
	    BufferedImage qrCodeImage = fullImage.getSubimage(x,y,width,height);
	    ImageIO.write(qrCodeImage, "png", screenshot);
            return qrCodeImage;
	}

**Reply by billychen2014** [reply]
yep, you are smart, we can produce the part png without appium ,let me have a try.

**Reply by Higgens75** [reply]
I capture the whole screen before with appium and than snip it with the funtion above.

File newBaseline = driver.getScreenshotAs(OutputType.FILE);
BufferedImage newBase=generateImageByCoords(newBaseline);
File outputfile = new File(baselineFilename);
ImageIO.write(newBase, "png", outputfile);

**Reply by Higgens75** [reply]
and you should use the similarity funtions, there you can get a result screen which shows the differences between both images.
with the findelementByImage you get only a NotFoundException which is not very helpfull.

**Reply by billychen2014** [reply]
hmm, i just did a test like what you said:

1. using driver.save_screenshot ==> got the full screenshot named **full.png**
2. using capture function(write by my self ) to get part.png
3. using driver.find_element_by_image("./part.png")

it looks like it always did not works .hmm, maybe python did not works well .

appium: 2.0.0-rc.4
testing device: iOS
python client

**Reply by Higgens75** [reply, marked as answer]
as i wrote, the notFoundException is not very helpfull in this scenario.
You should try the Image comparison commands from the first tutorial, where you can get some detailed output what has not worked or where the differences between both images are.

**Reply by billychen2014** [reply]
hmm, it looks like need to add one setting :  **"fixImageTemplateScale":True**

after doing this ,the finding function works well.

thanks a lot
